rust/arrow/src/ipc/file/reader.rs

```diff
@@ -0,0 +1,358 @@
+            let block = self.blocks[self.current_block];
+            self.current_block = self.current_block + 1;
+
+            // read length from end of offset
+            let meta_len = block.metaDataLength() - 4;
+
+            let mut block_data = vec![0; meta_len as usize];
+            self.reader
+                .seek(SeekFrom::Start(block.offset() as u64 + 4))?;
+            self.reader.read_exact(&mut block_data)?;
+
+            let c_block = ipc::get_root_as_message(&block_data[..]);
+
+            match c_block.header_type() {
+                ipc::MessageHeader::Schema => {
+                    panic!("Not expecting a schema when messages are read")
+                }
+                ipc::MessageHeader::DictionaryBatch => {
+                    unimplemented!("reading dictionary batches not yet supported")
+                }
+                ipc::MessageHeader::RecordBatch => {
+                    let c_batch = c_block.header_as_record_batch().unwrap();
+                    // read array data
+                    let mut a_data = vec![0; block.bodyLength() as usize];
+                    self.reader.seek(SeekFrom::Start(
+                        block.offset() as u64 + block.metaDataLength() as u64,
+                    ))?;
+                    self.reader.read_exact(&mut a_data)?;
+
+                    // construct buffers from their blocks
+                    let c_buffers = c_batch.buffers().unwrap();
+
+                    // get fields and determine number of buffers to use for each
+                    let c_nodes = c_batch.nodes().unwrap();
+                    let mut buffer_num = 0;
+                    let mut field_num = 0;
+                    let mut arrays = vec![];
+                    for c_node in c_nodes {
+                        let field = self.schema.field(field_num);
+                        let (array, buffer) = create_array(
+                            c_node,
+                            field.data_type(),
+                            &a_data,
+                            c_buffers,
+                            buffer_num,
+                        );
+                        field_num = field_num + 1;
+                        buffer_num = buffer;
+
+                        arrays.push(array);
+                    }
+
+                    RecordBatch::try_new(self.schema.clone(), arrays)
+                        .map(|batch| Some(batch))
+                }
+                ipc::MessageHeader::SparseTensor => panic!(),
+                ipc::MessageHeader::Tensor => panic!("Can't be Tensor"),
+                ipc::MessageHeader::NONE => panic!("Can't be NONE"),
+            }
+        } else {
+            Ok(None)
+        }
+    }
+
+    /// Read a specific record batch
+    ///
+    /// Sets the current block to the batch number, and reads the record batch at that
+    /// block
+    pub fn read_batch(&mut self, batch_num: usize) -> Result<Option<RecordBatch>> {
```

Comment:
nit: wondering if we can change this to something like `set_index` which just change the `current_block` and the caller then have to call `next()` to actually read the batch.

Otherwise, this does have a side effect of advancing the current block index, which may caught people by surprise. We should at least point this out in the method comments.

Reply:
Done